### PR TITLE
fees on some redemptions

### PR DESCRIPTION
**Pull request by mejango**
Two things:

- ~~There may be a low-severity bug with redemptions with delegate allocations where the full reclaimAmount is sent to the beneficiary, despite some funds perhaps allocated to the delegates beforehand. reclaimAmount should be inclusive of funds sent to delegates, therefor the amount to send to beneficiary should only be the difference. It's low severity because no one is using redemption delegate allocations yet in their treasury designs. Redemeption delegate allocations may not work until this is fixed.~~  Not the case. The accounting is correct as is.

- Added revenue source for JBX through redemptions. When someone redeems their project tokens to access funds from their treasury, a fee (same percentage as payout fee) is taken from the reclaimedAmount if and only if the project's redemption rate is not 100%. The _beneficiary of the reclaimed funds from the redemption would also be the beneficiary of the JBX. If a project is using redemption rate for 1:1 refunds, a fee will not be taken. The motivation is emerging treasury design prototypes that are more redemption focused than payout focused.

Importantly, no interfaces have been changed. The new `JBPayoutRedemptionPaymentTerminal3_1_1` adheres fully to `IJBPayoutRedemptionPaymentTerminal3_1`. The only change that integrators should be aware of is that the `RedeemTokens` event now includes this fee value.

### CODE REVIEW

File: contracts/abstract/JBPayoutRedemptionPaymentTerminal3_1_1.sol [hunk 1]
```diff
@@ -855,9 +857,7 @@ abstract contract JBPayoutRedemptionPaymentTerminal3_1_1 is
           _metadata
         );
 
-        uint256 _numDelegates = _delegateAllocations.length;
```
Comment (viraj124) [inline contracts/abstract/JBPayoutRedemptionPaymentTerminal3_1_1.sol]: we save gas with this wondering why was this removed

File: contracts/abstract/JBPayoutRedemptionPaymentTerminal3_1_1.sol [hunk 2]
```diff
@@ -0,0 +1,1803 @@
...
+      ) = IJBSingleTokenPaymentTerminalStore3_1_1(store).recordRedemptionFor(
+        _holder,
+        _projectId,
+        _tokenCount,
+        _memo,
+        _metadata
+      );
+
+      // Set the reference to the fee discount to apply. No fee if the beneficiary is feeless or if the redemption rate is at its max.
+      _feeDiscount = isFeelessAddress[_beneficiary] ||
```
Comment (drgorillamd) [inline contracts/abstract/JBPayoutRedemptionPaymentTerminal3_1_1.sol:833]: marking biggest changes for review - starting here: fee on redemption

File: contracts/abstract/JBPayoutRedemptionPaymentTerminal3_1_1.sol [hunk 3]
```diff
@@ -0,0 +1,1803 @@
...
+      uint256 _error;
+      bytes memory _reason;
+
+      if (
+        ERC165Checker.supportsInterface(
+          address(_split.allocator),
+          type(IJBSplitAllocator).interfaceId
+        )
+      )
+        // If this terminal's token is ETH, send it in msg.value.
```
Comment (drgorillamd) [inline contracts/abstract/JBPayoutRedemptionPaymentTerminal3_1_1.sol:1292]: marking biggest changes for review - starting here: revert refactor

File: contracts/abstract/JBPayoutRedemptionPaymentTerminal3_1_1.sol [hunk 4]
```diff
@@ -0,0 +1,1803 @@
...
+        netPayoutAmount,
+        decimals,
+        _projectId,
+        _group,
+        _split
+      );
+
+      // Trigger the allocator's `allocate` function.
+      uint256 _error;
+      bytes memory _reason;
```
Comment (drgorillamd) [inline contracts/abstract/JBPayoutRedemptionPaymentTerminal3_1_1.sol:1284]: can't we just use this one and assign accordingly?

File: contracts/abstract/JBPayoutRedemptionPaymentTerminal3_1_1.sol [hunk 5]
```diff
@@ -0,0 +1,1803 @@
...
+        _projectId,
+        _tokenCount,
+        _memo,
+        _metadata
+      );
+
+      // Set the reference to the fee discount to apply. No fee if the beneficiary is feeless or if the redemption rate is at its max.
+      _feeDiscount = isFeelessAddress[_beneficiary] ||
+        (_fundingCycle.redemptionRate() == JBConstants.MAX_REDEMPTION_RATE &&
+          _fundingCycle.ballotRedemptionRate() == JBConstants.MAX_REDEMPTION_RATE) ||
```
Comment (drgorillamd) [inline contracts/abstract/JBPayoutRedemptionPaymentTerminal3_1_1.sol:835]: Shouldn't this check the ballot is Active (eg `fundingCycleStore.currentBallotStateOf(_projectId);`)?

File: contracts/abstract/JBPayoutRedemptionPaymentTerminal3_1_1.sol [hunk 6]
```diff
@@ -0,0 +1,1803 @@
...
+      }
+    }
+
+    // Keep a reference to the fee taken from the reclaim amount.
+    uint256 _reclaimAmountFee;
+
+    // Send the reclaimed funds to the beneficiary.
+    if (reclaimAmount != 0) {
+      // Get the fee for the reclaimed amount.
+      _reclaimAmountFee = _feeAmount(reclaimAmount, fee, _feeDiscount);
```
Comment (drgorillamd) [inline contracts/abstract/JBPayoutRedemptionPaymentTerminal3_1_1.sol:922]: fee is potentially sloaded twice (ln 836) + if we fix by having it on stack here, why not testing fee == 0 and shortcut?

File: contracts/abstract/JBPayoutRedemptionPaymentTerminal3_1_1.sol [hunk 7]
```diff
@@ -0,0 +1,1803 @@
...
+        )
+      )
+        // If this terminal's token is ETH, send it in msg.value.
+        try
+          _split.allocator.allocate{value: token == JBTokens.ETH ? netPayoutAmount : 0}(_data)
+        {} catch (bytes memory __reason) {
+          _reason = __reason;
+          _error = 1;
+        }
+      else _error = 2;
```
Comment (drgorillamd) [inline contracts/abstract/JBPayoutRedemptionPaymentTerminal3_1_1.sol:1299]: `else _reason = abi.encode('IERC165 fail');`

File: test/jb_payment_terminal_3_1_1/migrate.test.js [hunk 8]
```diff
@@ -0,0 +1,274 @@
...
+      deployMockContract(deployer, jbErc20PaymentTerminal.abi),
+      deployMockContract(deployer, jbPaymentTerminalStore.abi),
+      deployMockContract(deployer, jbOperatoreStore.abi),
+      deployMockContract(deployer, jbProjects.abi),
+      deployMockContract(deployer, jbSplitsStore.abi),
+      deployMockContract(deployer, jbPrices.abi),
+      deployMockContract(deployer, jbToken.abi),
+    ]);
+
+    const jbCurrenciesFactory = await ethers.getContractFactory('JBCurrencies');
```
Comment (drgorillamd) [inline test/jb_payment_terminal_3_1_1/migrate.test.js:55]: maybe we should await them together too?

File: contracts/abstract/JBPayoutRedemptionPaymentTerminal3_1_1.sol (hunk 5 above)
Comment (mejango) [reply contracts/abstract/JBPayoutRedemptionPaymentTerminal3_1_1.sol:835]: it could, but thats a whole nother lookup. the local contract doesn't have a fundingCycleStore instance. im down to go this route, or just keep it so fees apply if either setting isn't 100%.

File: contracts/abstract/JBPayoutRedemptionPaymentTerminal3_1_1.sol [hunk 9]
```diff
@@ -0,0 +1,1803 @@
...
+        ERC165Checker.supportsInterface(
+          address(_split.allocator),
+          type(IJBSplitAllocator).interfaceId
+        )
+      )
+        // If this terminal's token is ETH, send it in msg.value.
+        try
+          _split.allocator.allocate{value: token == JBTokens.ETH ? netPayoutAmount : 0}(_data)
+        {} catch (bytes memory __reason) {
+          _reason = __reason;
```
[the remark this answers is not in the document]
Comment (mejango) [reply contracts/abstract/JBPayoutRedemptionPaymentTerminal3_1_1.sol:1296]: try it out! im curious

[the remark this answers is not in the document]
Comment (mejango) [reply contracts/abstract/JBPayoutRedemptionPaymentTerminal3_1_1.sol:1296]: done

File: contracts/abstract/JBPayoutRedemptionPaymentTerminal3_1_1.sol [hunk 10]
```diff
@@ -855,9 +857,7 @@ abstract contract JBPayoutRedemptionPaymentTerminal3_1_1 is
           _metadata
         );
 
-        uint256 _numDelegates = _delegateAllocations.length;
-
-        for (uint256 _i; _i < _numDelegates; ) {
+        for (uint256 _i; _i < _delegateAllocations.length; ) {
```
Comment (viraj124) [inline contracts/abstract/JBPayoutRedemptionPaymentTerminal3_1_1.sol:871]: we should cache length to save gas wondering why was this reverted

File: contracts/abstract/JBPayoutRedemptionPaymentTerminal3_1_1.sol [hunk 11]
```diff
@@ -0,0 +1,1718 @@
...
+    // Set the leftover amount to the initial amount.
+    leftoverAmount = _amount;
+    // The total percentage available to split
+    uint256 leftoverPercentage = JBConstants.SPLITS_TOTAL_PERCENT;
+
+    // Get a reference to the project's payout splits.
+    JBSplit[] memory _splits = splitsStore.splitsOf(_projectId, _domain, _group);
+
+    // Transfer between all splits.
+    for (uint256 _i; _i < _splits.length; ) {
```
Comment (viraj124) [inline contracts/abstract/JBPayoutRedemptionPaymentTerminal3_1_1.sol:1190]: same here too

Comment (mejango) [reply contracts/abstract/JBPayoutRedemptionPaymentTerminal3_1_1.sol:1223]: this is a stack too deep thing. we could do via-ir, worth it?

File: contracts/abstract/JBPayoutRedemptionPaymentTerminal3_1_1.sol [hunk 12]
```diff
@@ -1233,12 +1231,7 @@ abstract contract JBPayoutRedemptionPaymentTerminal3_1_1 is
       _split = _splits[_i];
 
       // The amount to send towards the split.
-      uint256 _payoutAmount = _split.percent == leftoverPercentage
-        ? leftoverAmount
-        : PRBMath.mulDiv(_amount, _split.percent, JBConstants.SPLITS_TOTAL_PERCENT);
-
-      // Decrement the leftover percentage.
-      leftoverPercentage -= _split.percent;
+      uint256 _payoutAmount = PRBMath.mulDiv(_amount, _split.percent, _leftoverPercentage);
```
Comment (xBA5ED) [inline contracts/abstract/JBPayoutRedemptionPaymentTerminal3_1_1.sol:1234]: Is this correct? isn't it supposed to be:
```
PRBMath.mulDiv(_amount, _split.percent, JBConstants.SPLITS_TOTAL_PERCENT)
```

File: contracts/abstract/JBPayoutRedemptionPaymentTerminal3_1_1.sol [hunk 13]
```diff
@@ -0,0 +1,1835 @@
...
+  function _cancelTransferTo(address _to, uint256 _amount) internal virtual {
+    _to; // Prevents unused var compiler and natspec complaints.
+    _amount; // Prevents unused var compiler and natspec complaints.
+  }
+
+  /** 
+    @notice 
+    Verifies this terminal is a terminal of provided project ID.
+  */
+  function isTerminalOf(uint256 _projectId) internal view {
```
Comment (xBA5ED) [inline contracts/abstract/JBPayoutRedemptionPaymentTerminal3_1_1.sol:779]:
```suggestion
  function _isTerminalOf(uint256 _projectId) internal view {
```

File: contracts/abstract/JBPayoutRedemptionPaymentTerminal3_1_1.sol [hunk 14]
```diff
@@ -0,0 +1,1835 @@
...
+    address _beneficiary,
+    uint256 _minReturnedTokens,
+    bool _preferClaimedTokens,
+    string calldata _memo,
+    bytes calldata _metadata
+  ) external payable virtual override  returns (uint256) {
+    _token; // Prevents unused var compiler and natspec complaints.
+
+    // valid terminal check
+    isTerminalOf(_projectId);
```
Comment (xBA5ED) [inline contracts/abstract/JBPayoutRedemptionPaymentTerminal3_1_1.sol:346]:
```suggestion
    _isTerminalOf(_projectId);
```

File: contracts/abstract/JBPayoutRedemptionPaymentTerminal3_1_1.sol (hunk 12 above)
Comment (mejango) [reply contracts/abstract/JBPayoutRedemptionPaymentTerminal3_1_1.sol:1234]: i adapted it to remove a dependency on `leftoverAmount`. i think this works.

Comment (mejango) [reply contracts/abstract/JBPayoutRedemptionPaymentTerminal3_1_1.sol:1234]: the denominator _leftoverPercentage decreases as _amount decreases. When _split.percent == _leftoverPercent, only the _amount remains. same as before, just more direct terms. maybe there's a hole in this strategy tho

**Review by drgorillamd — APPROVED** [review]
